src/backend/cuda/kernel/reduce_by_key.hpp

```diff
@@ -0,0 +1,732 @@
+
+    // space to hold left/right-most keys of each reduced warp to check if
+    // reduction should happen accros boundaries
+    __shared__ Tk warpReduceLeftBoundaryKeys[nWarps];
+    __shared__ Tk warpReduceRightBoundaryKeys[nWarps];
+
+    // space to hold right-most values of each reduced warp to check if
+    // reduction should happen accros boundaries
+    __shared__ compute_t<To> warpReduceRightBoundaryVals[nWarps];
+
+    // space to compact and finalize all reductions within block
+    __shared__ Tk warpReduceKeysSmemFinal[nWarps * maxResPerWarp];
+    __shared__ compute_t<To> warpReduceValsSmemFinal[nWarps * maxResPerWarp];
+
+    //
+    // will hold final number of reduced elements in block
+    __shared__ int reducedBlockSize;
+    __shared__ int dim_ordering[4];
+
+    compute_t<To> init = Binary<compute_t<To>, op>::init();
+
+    if (threadIdx.x == 0) {
+        reducedBlockSize = 0;
+        int d            = 1;
+        dim_ordering[0]  = dim;
+        for (int i = 0; i < 4; ++i) {
+            if (i != dim) dim_ordering[d++] = i;
+        }
+    }
+    if (threadIdx.x < nWarps * maxResPerWarp)
+        warpReduceValsSmemFinal[threadIdx.x] = init;
+    __syncthreads();
+
+    Binary<compute_t<To>, op> reduce;
+    Transform<compute_t<Ti>, compute_t<To>, op> transform;
+
+    // load keys and values to threads
+    Tk k;
+    compute_t<To> v;
+    if (tidx < n) {
+        const int tid = bidw * vals.strides[dim_ordering[3]] +
+                        bidz * vals.strides[dim_ordering[2]] +
+                        bidy * vals.strides[dim_ordering[1]] +
+                        tidx * vals.strides[dim];  // index for batched inputs
+
+        k = keys.ptr[tidx];
+        v = transform(compute_t<Ti>(vals.ptr[tid]));
+        if (change_nan) v = IS_NAN(v) ? compute_t<To>(nanval) : v;
+    } else {
+        v = init;
+    }
+
+    Tk eq_check = (k != shfl_up_sync(0xFFFFFFFF, k, 1));
+    // mark threads containing unique keys
+    char unique_flag = (eq_check || (laneid == 0)) && (tidx < n);
+
+    // scan unique flags to enumerate unique keys
+    char unique_id = unique_flag;
+#pragma unroll
+    for (int offset = 1; offset < 32; offset <<= 1) {
+        char y = shfl_up_sync(0xFFFFFFFF, unique_id, offset);
+        if (laneid >= offset) unique_id += y;
+    }
+
+    //
+    // Reduce each warp by key
+    char all_eq = (k == shfl_down_sync(0xFFFFFFFF, k, 1));
+    if (all_sync(0xFFFFFFFF,
+                 all_eq)) {  // check special case of single key per warp
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 1));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 2));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 4));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 8));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 16));
+    } else {
+        // preform reduction for each of the unique keys
+        int eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, 1));
```

Comment:
Same as the other non-dim version function, this else code block can be condensed into a `#pragma unroll` based for-loop.